drivers/gpio/gpio_gecko.c

```diff
@@ -0,0 +1,425 @@
+#define NUMBER_OF_PORTS (member_size(GPIO_TypeDef, P) / \
+			 member_size(GPIO_TypeDef, P[0]))
+
+struct gpio_gecko_common_config {
+};
+
+struct gpio_gecko_common_data {
+	/* a list of all ports */
+	struct device *ports[NUMBER_OF_PORTS];
+	size_t count;
+};
+
+struct gpio_gecko_config {
+	GPIO_P_TypeDef *gpio_base;
+	GPIO_Port_TypeDef gpio_index;
+};
+
+struct gpio_gecko_data {
+	/* port ISR callback routine address */
+	sys_slist_t callbacks;
+	/* pin callback routine enable flags, by pin number */
+	u32_t pin_callback_enables;
+};
+
+static inline void gpio_gecko_add_port(struct gpio_gecko_common_data *data,
+				       struct device *dev)
+{
+	__ASSERT(dev, "No port device!");
+	data->ports[data->count++] = dev;
+}
+
+
+static int gpio_gecko_configure(struct device *dev,
+				int access_op, u32_t pin, int flags)
+{
+	const struct gpio_gecko_config *config = dev->config->config_info;
+	GPIO_P_TypeDef *gpio_base = config->gpio_base;
+	GPIO_Port_TypeDef gpio_index = config->gpio_index;
+	GPIO_Mode_TypeDef mode;
+	unsigned int out = 0;
+	bool interrupt_rising_edge = false;
+	bool interrupt_falling_edge = false;
+	bool enable_interrupt = false;
+
+	/* Check for an invalid pin configuration */
+	if ((flags & GPIO_INT) && (flags & GPIO_DIR_OUT)) {
+		return -EINVAL;
+	}
+
+	/* Interrupt on static level is not supported by the hardware */
+	if ((flags & GPIO_INT) && ((flags & GPIO_INT_EDGE) == GPIO_INT_LEVEL)) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
+
+		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+			mode = gpioModeInputPull;
+			out = 1; /* pull-up*/
+		} else if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
+			mode = gpioModeInputPull;
+			/* out = 0 means pull-down*/
+		}
+
+		mode = gpioModeInput;
+	} else { /* GPIO_DIR_OUT */
+		mode = gpioModePushPull;
+	}
+	/* The flags contain options that require touching registers in the
+	 * GPIO module and the corresponding PORT module.
+	 *
+	 * Start with the GPIO module and set up the pin direction register.
+	 * 0 - pin is input, 1 - pin is output
+	 */
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		GPIO_PinModeSet(gpio_index, pin, mode, out);
+	} else {	/* GPIO_ACCESS_BY_PORT */
+		gpio_base->MODEL = (mode << 28) | (mode << 24) | (mode << 20)
+			| (mode << 16) | (mode << 12) | (mode << 8)
+			| (mode << 4) | mode;
+		gpio_base->MODEH = (mode << 28) | (mode << 24) | (mode << 20)
+			| (mode << 16) | (mode << 12) | (mode << 8)
+			| (mode << 4) | mode;
+		gpio_base->DOUT = (out ? 0xFFFF : 0x0000);
+	}
+
+	if (flags & GPIO_INT) {
+		enable_interrupt = true;
+		if (flags & GPIO_INT_EDGE) {
+			if (flags & GPIO_INT_ACTIVE_HIGH) {
+				interrupt_rising_edge = true;
+			} else if (flags & GPIO_INT_DOUBLE_EDGE) {
+				interrupt_rising_edge = true;
+				interrupt_falling_edge = true;
+			} else {
+				interrupt_falling_edge = true;
+			}
+		} else { /* GPIO_INT_LEVEL */
+			 /* should never get here, this is already handled */
+			return -ENOTSUP;
```

Comment:
yeah so this else is totally useless then...

Reply:
done